docs/architecture.md

```diff
@@ -0,0 +1,144 @@
+  Architecture of Seccomp Agent.
+---
+
+
+## From Kubernetes to Seccomp Agent
+
+![From Kubernetes to Seccomp Agent](kubernetes-to-seccomp-agent.png "From Kubernetes to Seccomp Agent")
+
+Steps:
+- A user creates a pod with `kubectl apply -f pod.yaml`.
+  The file `pod.yaml` specifies the seccomp profile `foo.json` as follows:
+  ```
+  apiVersion: v1
+  kind: Pod
+  metadata:
+    name: mypod
+  spec:
+    securityContext:
+      seccompProfile:
+        type: Localhost
+        localhostProfile: foo.json
+  ```
+- Once the pod is scheduled on a node, a Kubelet on a node receives the pod
+  specification.
+- The Kubelet calls the method
+  [StartContainer](https://github.com/kubernetes/cri-api/blob/v0.22.0-alpha.2/pkg/apis/runtime/v1/api.proto#L65)
+  on containerd using the
+  [Container Runtime Interface (CRI)](https://github.com/kubernetes/cri-api).
+  The method includes the field `SeccompProfilePath` with the value
+  `/var/lib/kubelet/seccomp/foo.json`.
+- Containerd reads the file `/var/lib/kubelet/seccomp/foo.json`:
+  ```
+  {
+    "listenerPath": "/run/seccomp-agent.socket",
+    "defaultAction": "SCMP_ACT_ALLOW",
+    "syscalls": [
+      {
+        "names": ["mount"],
+        "action": "SCMP_ACT_NOTIFY"
+      }
+  }
+  ```
+- Containerd prepares config.json that includes the file foo.json from above:
+  ```
+  {
+    "linux": {
+      "seccomp": {
+        "listenerPath": "/run/seccomp-agent.socket",
+        "defaultAction": "SCMP_ACT_ALLOW",
+        "syscalls": [
+          {
+            "names": ["mount"],
+            "action": "SCMP_ACT_NOTIFY"
+          }
+  ```
+- Containerd starts runc with `config.json`.
+- Runc gets the seccomp file descriptor by calling the seccomp() system call
+  via libseccomp-golang.
+- Runc sends the seccomp file descriptor to the seccomp agent, along with
+  following
+  [Container Process State](https://github.com/opencontainers/runtime-spec/blob/1c3f411f041711bbeecf35ff7e93461ea6789220/config-linux.md#containerprocessstate):
+  ```
+  {
+    "ociVersion": "0.2.0",
+    "fds": ["seccompFd"],
+    "pid": 4422,
+    "metadata": "",
+    "state": {
+      "ociVersion": "0.2.0",
+      "id": "8eea9f6d...",
+      "status": "creating",
+      "pid": 4422,
+      "bundle": "rootfs",
+      "annotations": {}
+    }
+  }
+  ```
+- At this point, the Seccomp Agent has the seccomp file descriptor and the
+  Container Process State contains the following
+  [annotations](https://github.com/containerd/containerd/blob/v1.5.2/pkg/cri/annotations/annotations.go)
+  to identify the pod. Additional information such as the ServiceAccount could
+  be queried to the Kubernetes API.
+  - Pod namespace: `io.kubernetes.cri.sandbox-namespace`
+  - Pod name: `io.kubernetes.cri.sandbox-name`
+  - Container name: `io.kubernetes.cri.container-name`
+
+## Example of system call performed on behalf of the process
+
+![Mount syscall with Seccomp Notify](example-syscall-mount.png "Mount syscall with Seccomp Notify")
```

Comment:
Can we give more context (like unprivileged builds)? This is something a user almost never does, it is not clear why we are talking about this, IMHO